Taras samchuk hw5

Sam4uk opened the pull request:
Реалізація однозвязного списку

Sam4uk committed:
node_one_ptr

Sam4uk committed:
---

Sam4uk committed:
ReWriteAll

Sam4uk committed:
публікування

MykolaSuperman commented on Homework5/makefile:
```diff
@@ -0,0 +1,43 @@
+# формула імен о-файлів
+OBJECTS=$(SOURCES:.cpp=.o)
+# ім'я програми
+EXECUTABLE=Taras_Samchuk_HW5.bin
```
bin??? я думаю что это вы зря, так как лучше такое имя не давать. На разных платформах может быть разное поведение. и чтобы исполняемый файл был как bin это я не видел

MykolaSuperman commented on Homework5/Classic_TS_HW5.hpp:
```diff
@@ -0,0 +1,13 @@
+void pop_front(node_one_ptr **root);
+void clear_list(node_one_ptr **root);
+int view_front(node_one_ptr *root);
+int view_back(node_one_ptr *root);
```
лучше место `view_*` написать без `view`, так будет намного наглядней

MykolaSuperman commented on Homework5/Classic_TS_HW5.hpp:
```diff
@@ -0,0 +1,13 @@
+  int Data{};
+  node_one_ptr *next{nullptr};
+};
+
```
я бы добавил еще пару методов
```
node_one_ptr* insert (node_one_ptr* pos, int  value);
node_one_ptr* find (int  value);
```

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+
+void clear_list(node_one_ptr *root) {
+  node_one_ptr *arrow = root;
+  root->next = nullptr;
```
вообще не понял)
а цикл ниже будет выполняться?

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+  delete arrow;
+};
+
+void pop_front(node_one_ptr **root) {
```
можно чтобы без двойного указателя писать `*&` это более С++ стиль, тем самым вам в коде не нужно использовать лишнее разыменование

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+    arrow = arrow->next;
+  }
+  step_back->next = nullptr;
+  delete arrow;
```
ну вопервых если у меня `root ` состоит с одной ноды, как у меня обнулиться сам `root`.
Потом если у вас одна нода, тогда у вас цикл `while` не сработает и у вас выражение `step_back->next = nullptr;` приведет к неопределенному поведению.

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+};
+
+void push_front(node_one_ptr **root, const int info) {
+  node_one_ptr *NewNode = new node_one_ptr;
```
а если так 
```
node_one_ptr *NewNode = new node_one_ptr{info, *root};
```

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+void push_front(node_one_ptr **root, const int info) {
+  node_one_ptr *NewNode = new node_one_ptr;
+  NewNode->Data = info;
+  if (root == nullptr) {
```
а не `*root == nullptr` ?

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+    *root = NewNode;
+    return;
+  }
+  NewNode->next = *root;
```
это не нужно если вы например примените выше форму инициализации

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+  }
+  NewNode->next = *root;
+  *root = NewNode;
+};
```
если написать, наверное, форму инициализации что я описал выше, то вам наверное и оператор `if` не нужен

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+void push_back(node_one_ptr **root, const int info) {
+  node_one_ptr *NewNode = new node_one_ptr;
+  NewNode->Data = info;
+  if (root == nullptr) {
```
`*root` ?

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,98 @@
+    arrow = arrow->next;
+  }
+  std::cout << k_mes_nullptr << std::endl;
+  return;
```
а зачем нужен return?

Sam4uk replied:
Задум був щоб цей метод повертав `true` якщо відбулося виведення і `false` якщо нічого не виводилося (порожньо).

MykolaSuperman replied:
Функция печати возращала true или false, это было просто шокируеще))

Sam4uk committed:
Update makefile

Sam4uk committed:
---

Sam4uk committed:
-

Sam4uk committed:
Застосування зауважень та додаванння методів
node_one_ptr *insert(node_one_ptr *&head, node_one_ptr *pos, const int value);
node_one_ptr *find(node_one_ptr *head, const int value);

Sam4uk committed:
UpLOAD

Sam4uk commented:
Застосовані зауваження та додані методи 

node_one_ptr *insert(node_one_ptr *&head, node_one_ptr *pos, const int value);
node_one_ptr *find(node_one_ptr *head, const int value);

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+    std::cout << k_mes_nullptr << std::endl;
+    return;
+  }
+  node_one_ptr *arrow = head;
```
а зачем тут лишний указатель `arrow`? если вы можете использовать `head` у вас же сам указатель `head` то не измениться?

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+  }
+  node_one_ptr *step_back{nullptr};
+  node_one_ptr *arrow = head;
+  while (arrow->next != nullptr) {
```
а если не использовать `step_back` ?
а например написать так 
```
arrow->next->next != nullptr
```
и ниже 
```
delete arrow->next;
arrow->next = nullptr;
```
как вы думаете так лучше будет? и будет ли вообще работать?

Sam4uk replied:
Працювати буде в обох варіантах.
В моєму варіанті я займаю пам'ять під тимчасовий вказівник.
Ваш ліпший - не використовуєьбся додаткових ресурсів.

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+  node_one_ptr *delete_this = head;
+  node_one_ptr *hold = delete_this->next;
+  delete delete_this;
+  *head = *hold;
```
этой строки вообще не понял? зачем вам присваивать содержимое нод?
почему не 
```
head = hold;
```

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+
+void clear_list(node_one_ptr *&head) {
+  node_one_ptr *arrow = head;
+  head = nullptr;
```
такое уже делают всегда в конце функции, а не в начале

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+
+int viewfront(node_one_ptr *head) {
+  if (head == nullptr) {
+    return 0;
```
конечно выглядит не очень) т.е. ноль это не валидное число?

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+  if (head == nullptr) {
+    return {};
+  }
+  node_one_ptr *arrow = head;
```
зачем вам лишний временный указатель

Sam4uk replied:
Коли параметром функції є вказівник, я починаю боятися нашкодити йому (щоб не змінити вхідне значення)

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,128 @@
+    return nullptr;
+  }
+  node_one_ptr *NewNode = new node_one_ptr{value, nullptr};
+  if (head == nullptr) {
+    head = NewNode;
+    return head;
+  }
```
я думаю что проверка head в этом коде наверное не меет смыл, т.к. pos  есть часть вашего head

Sam4uk committed:
Застосування зауважень (6 квітня)

Sam4uk committed:
Fix

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,119 @@
+static const char k_mes_head[]{"}>"};
+void print_list(node_one_ptr *head) {
+  std::cout << k_mes_head;
+  if (head == nullptr) {
+    std::cout << k_mes_nullptr << std::endl;
+    return;
+  }
+  while (head != nullptr) {
+    std::cout << head->Data << k_mes_arrow;
+    head = head->next;
+  }
```
```suggestion
    while (head != nullptr) {
      std::cout << head->Data << k_mes_arrow;
      head = head->next;
    }
```

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,119 @@
+  if (head == nullptr) {
+    return;
+  }
+  node_one_ptr *delete_this = head;
+  node_one_ptr *hold = delete_this->next;
+  delete delete_this;
+  head = hold;
```
```suggestion
  node_one_ptr *hold = head->next;
  delete head;
  head = hold;
```

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,119 @@
+};
+
+void clear_list(node_one_ptr *&head) {
+  node_one_ptr *arrow = head;
```
вам тут промежуточный указатель не нужен. У вас и так выходной head должен быть nullptr поэтому не вижу смысла использовать промежуточный указатель

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,119 @@
+  if (head == nullptr) {
+    return {};
+  }
+  while (head->next != nullptr) {
+    head = head->next;
+  };
```
у вас часто этот код повторяеться может написать для него типа get_last_element

MykolaSuperman commented on Homework5/Classic_TS_HW5.cpp:
```diff
@@ -0,0 +1,119 @@
+  node_one_ptr *arrow = head;
+  while (arrow->next != nullptr && arrow != pos) {
+    arrow = arrow->next;
+  }
```
я думаю что если мы дошли до arrow->next != nullptr то наверное не стоит добавлять элемент в конец списка

Sam4uk replied:
Не розумію чому не варто додавати

Sam4uk committed:
Update Homework5/Classic_TS_HW5.cpp

Co-Authored-By: MykolaSolyanko <solianko.nik@gmail.com>

Sam4uk committed:
upd Classic_TS_HW5.cpp
